Make Argon2.hash parameters configurable to allow for stronger user password hashes.

TonyLovesDevOps opened the pull request:
Regarding discussion #7670 with @rijkvanzanten

1.  Create generatePasswordHash util function to standardize how user passwords are hashed throughout the API instead of directly calling argon2.hash directly;
2.  Add configuration parameters to modify the behavior of the argon2.hash function used for generating user password hashes;
3.  Add docs and placeholders in api/example.env for new parameters;
4.  Update argon2 to latest release.

**Open Questions**

1. Does `PayloadService.hash` get called for anything _other than_ hashing user passwords? I didn't see any calls to it for other uses but I didn't test extensively;
2. There are calls to `argon2.hash` in [packages/sdk/tests/handlers/utils.test.ts](https://github.com/directus/directus/blob/acd41eb0be90933db09af6b02d3ffe0c56a64efd/packages/sdk/tests/handlers/utils.test.ts#L33-L34), [api/src/controllers/utils.ts](https://github.com/directus/directus/blob/d64ca14348184e69e4348c8c61bbf88ffcc9937a/api/src/controllers/utils.ts#L34), and [api/src/services/graphql.ts](https://github.com/directus/directus/blob/d64ca14348184e69e4348c8c61bbf88ffcc9937a/api/src/services/graphql.ts#L1607) -- these appear to not be user-password related so I did not update them. I want to make sure my assumption is correct and that we want to leave them as they are.

TonyLovesDevOps committed:
1.  Create generatePasswordHash util function to standardize how user
        passwords are hashed throughout the API instead of directly calling
        argon2.hash directly;
    2.  Add configuration parameters to modify the behavior of the argon2.hash
        function used for generating user password hashes;
    3.  Add docs and placeholders in api/example.env for new parameters;
    4.  Update argon2 to latest release.

TonyLovesDevOps committed:
1.  Create generatePasswordHash util function to standardize how user
        passwords are hashed throughout the API instead of directly calling
        argon2.hash directly;
    2.  Add configuration parameters to modify the behavior of the argon2.hash
        function used for generating user password hashes;
    3.  Add docs and placeholders in api/example.env for new parameters;
    4.  Update argon2 to latest release.

TonyLovesDevOps committed:
Also update CLI init command to call generatePasswordHash vs. calling argon2.hash directly.

TonyLovesDevOps committed:
Also update CLI init command to call generatePasswordHash vs. calling argon2.hash directly.

joselcvarela committed:
docs: sdk auth.password.request url argument (#7757)

aidenfoxx commented:
It might be a good idea to add the hash algorithm to the `shared` package since there are a lot of places in Directus that assume a default `argon2` configuration, including the `cli` package and soon to be #6942.

rijkvanzanten committed:
Show a warning if PostGIS is missing (#7759)

* Show a warning if PostGIS is missing

* Remove sqlite-extensions flag

* Remove unused import

benhaynes committed:
Update WYSIWYG styling

joselcvarela committed:
translations: fix user.status_invited (#7760)

rijkvanzanten committed:
New Crowdin updates (#7739)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Czech)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Russian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Czech)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

rijkvanzanten committed:
Fix postgis check

rijkvanzanten committed:
Fix branch emitter logic from grand-to-child (#7763)

rijkvanzanten committed:
New Crowdin updates (#7762)

* New translations en-US.yaml (Russian)

* New translations en-US.yaml (Slovenian)

* New translations en-US.yaml (Chinese Simplified)

* New translations en-US.yaml (French)

* New translations en-US.yaml (Arabic)

* New translations en-US.yaml (Catalan)

* New translations en-US.yaml (German)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (German)

* New translations en-US.yaml (Hungarian)

rijkvanzanten committed:
Fix tree-select not allowing groups to be opened in selection mode

rijkvanzanten committed:
Optimize tree-select open state

Nitwel committed:
Warn the user when a collapsed group field had an error (#7738)

* warn the user when a collapsed group field had an error

* Replace icon

* Reduce icon size

Co-authored-by: rijkvanzanten <rijkvanzanten@me.com>

rijkvanzanten committed:
New translations en-US.yaml (Hungarian) (#7764)

azrikahar committed:
fix WYSIWYG field focus event (#7756)

renovate[bot] committed:
Update vue monorepo to v3.2.7 (#7766)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

rijkvanzanten commented on docs/reference/environment-variables.md:
```diff
@@ -73,6 +73,38 @@ your project and API on different domains, make sure to verify your configuratio
+1. Modifying `HASH_MEMORY_COST` and/or `HASH_PARALLELISM` will affect the amount of memory directus uses when computing
+   hashes; each thread gets `HASH_MEMORY_COST` amount of memory, so the total additional memory will be these two values
+   multiplied. This may cause out of memory errors, especially when running in containerized environments;
+2. Unless you absolutely know what you are doing, you should not override the default settings for `HASH_VERSION`,
+   `HASH_SALT`, and `HASH_SALT_LENGTH`. Setting `HASH_SALT` will use the **same salt for all user password hashes**
+   which is almost certainly not what you want. Setting `HASH_VERSION` and `HASH_SALT_LENGTH` are specifically advised
+   against on the [node-argon2 library options page](https://github.com/ranisalt/node-argon2/wiki/Options).
```
I kinda feel like we should omit these from the docs altogether, in favor of a note like "Every other env var is passed on to argon2." similarly how we do it for the database config stuff. I'd rather not present the user an option which they are not supposed to use unless they absolutely know what they're doing.

TonyLovesDevOps replied:
Sure, I can remove the parameters I put in warnings about from being called out specifically, and copy the verbiage from the DB config parameter section. I think it would be useful to include warnings about memory usage - what do you think about that?

rijkvanzanten replied:
> I think it would be useful to include warnings about memory usage - what do you think about that?

Yes totally 👍🏻

rijkvanzanten reviewed (changes requested):
Right now, argon2.hash is used for 2 main purposes: Generating password hashes for directus_users, and generating hashes for the generic "hash" field type. I believe both should be able to be configured in this fashion. Should we have a single hash configuration, and use it for both, or have two sets of config?

rijkvanzanten committed:
Fix hash link in docs module (#7768)

renovate[bot] committed:
Update dependency knex-schema-inspector to v1.6.0 (#7769)

oreilles committed:
Use OpenMapTiles font instead of ArcGIS (#7780)

definiteIymaybe committed:
tiny rewrite in docs/reference/filter-rules (#7771)

renovate[bot] committed:
Update dependency sass to v1.39.0 (#7770)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

TonyLovesDevOps commented:
> Right now, argon2.hash is used for 2 main purposes: Generating password hashes for directus_users, and generating hashes for the generic "hash" field type. I believe both should be able to be configured in this fashion. Should we have a single hash configuration, and use it for both, or have two sets of config?

If we want to support modifying the parameters for hashing non-passwords, I think we'd want those to be separate config sets. The reason being is that I expect our desired password hash settings to cause the `argon2.hash` call to take 1-2 seconds and use ~200MiB of memory for each password hash operation, and I wouldn't want that for just generating non-password hashes.

What about making changes like:

1. Prepend the second set of env vars with `PASSWORD_` to the current parameter names e.g. `PASSWORD_HASH_TIME_COST` and use those in the `generatePasswordHash` function;
2. Create a new `generateHash` function in its own `utils` function/file and use the existing named params for that.

rijkvanzanten commented:
> What about making changes like:
> 
> Prepend the second set of env vars with PASSWORD_ to the current parameter names e.g. PASSWORD_HASH_TIME_COST and use those in the generatePasswordHash function;
> Create a new generateHash function in its own utils function/file and use the existing named params for that.


That sounds perfect to me! Thanks 🙂

Commit:
Expose logger through ExtensionContext (#7777)

Fixes #7737.

oreilles committed:
Fixes bug when trying to edit geometry in code interface. (#7778)

renovate[bot] committed:
Update vue monorepo to v3.2.8 (#7785)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency npm to v7.22.0 (#7786)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

benhaynes committed:
Improve card selection styling (#7787)

rijkvanzanten committed:
Add environment variable to force-exclude tables from Directus (#7789)

rijkvanzanten committed:
New Crowdin updates (#7765)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Hungarian)

* New translations en-US.yaml (Polish)

rijkvanzanten committed:
v9.0.0-rc.92

rijkvanzanten committed:
Update changelog.md

smilledge committed:
Update logging in CLI commands (#7676)

* Replace console logs with logger in CLI commands

* Fix double sparkles in migration command

* Apply suggestions from code review

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

* Use stdout for init command output

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

azrikahar committed:
Fix WYSIWYG field not being cleared after "Save and Create New" (#7754)

* fix WYSIWYG field not being cleared

* return empty string when null for tinymce

* Update app/src/components/v-form/form-field.vue

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

Nitwel committed:
fix wrong icon (#7800)

Nitwel committed:
fix false is proper value (#7796)

licitdev committed:
Fix incorrect database exclude tables environment variable in docs. (#7798)

Commit:
Datetime interface "Set to now" sets seconds to 0 (#7794)

When "Include seconds" is disabled.

Fixes #7779

oreilles committed:
Use https for openmaptiles fonts. (#7801)

azrikahar committed:
fix type for deep query with underscore prefix (#7815)

oreilles committed:
Map selection behaviour (#7811)

* Use https for openmaptiles fonts.

* Changed map selection behaviour: replace instead of adding them by default.

oreilles committed:
Fix map interface controls not showing (#7812)

joselcvarela committed:
Fix return error for GraphQL mutations (#7814)

joselcvarela committed:
Apply "in" to query even if array is empty (#7816)

* fetch only avatar id for current user

* apply "in" to query even if array is empty
if the "in" array is empty we need to add it to query anyways,
otherwise all records will be attached causing out of memory on
subsequent calls

renovate[bot] committed:
Update dependency pinia to v2.0.0-rc.7 (#7817)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

HarunKilic committed:
Update modules.md (#7795)

TonyLovesDevOps commented:
> Right now, argon2.hash is used for 2 main purposes: Generating password hashes for directus_users, and generating hashes for the generic "hash" field type. I believe both should be able to be configured in this fashion. Should we have a single hash configuration, and use it for both, or have two sets of config?

It's a good thing you asked about this. I found that the answer to my own question of `Does PayloadService.hash get called for anything other than hashing user passwords?` with a "Yes - every time an item is added to a collection with a `Hash` field type"!

Is there any way to know in [the hash method in PayloadService](https://github.com/directus/directus/blob/main/api/src/services/payload.ts#L49-L57) what the context of the call is (i.e. if it's generating a user password or not)?

A few options I can think of:
1. Stick with only one set of `HASH_...` configuration parameters and use it for both passwords and hash field values, and add a note to the env var docs letting people know that it will affect all calls to generate hashes;
2. Find some way to differentiate between calls to hash user passwords and calls to generate a hash for the Hash field type and use the `PASSWORD_HASH_...` and `HASH_...` parameters as appropriate;
3. Something I haven't thought of... cry, maybe? 😭

rijkvanzanten commented:
Option 1 would certainly be the easiest to implement, maintain, and document 🤔 

Why don't we start with that, as it should allow enough flexility for the immediate use case to be satisfied. If the specific need arises to configure the hashing for users and regular hashes separately, we can go through the regular feature request flow to make that happen (it would be an addition, not a breaking change in the future0

azrikahar committed:
make csv imports try to parse values to json (#7820)

azrikahar committed:
fix minor typo in api-hooks.md (#7821)

renovate[bot] committed:
Update dependency knex to v0.95.11 (#7819)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

TonyLovesDevOps committed:
Rename hash utils file/function to be not password-specific, update all calls to argon2.hash to use the new function, update docs.

TonyLovesDevOps committed:
Rename hash utils file/function to be not password-specific, update all calls to argon2.hash to use the new function, update docs.

TonyLovesDevOps committed:
Merge branch 'directus-main'

TonyLovesDevOps commented:
I *think* I've addressed all the changes that were requested. I also rebased my fork so it added all of the commits since I opened the PR... 🙄 let me know if you want me to close this PR and open a new one with just my commits.

TonyLovesDevOps committed:
Preserve old behavior of hash function in PayloadService, in case 'value' is not a String.

rijkvanzanten committed:
Tweak docs

rijkvanzanten reviewed (approved):
LGTM!

TonyLovesDevOps commented:
🎉 Thanks for the guidance and for getting this merged quickly!

TonyLovesDevOps commented on api/src/services/users.ts:
```diff
@@ -339,7 +339,7 @@ export class UsersService extends ItemsService {
 		}
 
-		const passwordHashed = await argon2.hash(password);
+		const passwordHashed = generateHash(password);
```
Bob Saget! I forgot an `await` here. :hurtrealbad:

rijkvanzanten replied:
Happens to the best of us! It was already reported, and then fixed by @aidenfoxx ☺️
